File: scripts/import_inventory.py
"""
Script to import inventory items from Excel file to Supabase database.
"""
import pandas as pd
import httpx
import os
import uuid
from datetime import datetime

# Supabase configuration
SUPABASE_URL = "https://edumcnnilpnbdxcjpchw.supabase.co"
SUPABASE_KEY = "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJzdXBhYmFzZSIsInJlZiI6ImVkdW1jbm5pbHBuYmR4Y2pwY2h3Iiwicm9sZSI6ImFub24iLCJpYXQiOjE3NjczMzI1ODEsImV4cCI6MjA4MjkwODU4MX0.nRPDvserHqLnNx78UArKHuzJp5J4C0-FT9noHtmXfFU"

headers = {
    "apikey": SUPABASE_KEY,
    "Authorization": f"Bearer {SUPABASE_KEY}",
    "Content-Type": "application/json",
    "Prefer": "return=representation"
}

def get_departments():
    """Get all departments"""
    response = httpx.get(
        f"{SUPABASE_URL}/rest/v1/departments",
        headers=headers
    )
    print(f"Departments response status: {response.status_code}")
    print(f"Departments response: {response.text[:500]}")
    return response.json()

def get_classifications(department_id):
    """Get classifications for a department"""
    response = httpx.get(
        f"{SUPABASE_URL}/rest/v1/warehouse_classifications",
        headers=headers,
        params={"department_id": f"eq.{department_id}"}
    )
    return response.json()

def get_locations(classification_id):
    """Get locations for a classification"""
    response = httpx.get(
        f"{SUPABASE_URL}/rest/v1/warehouse_locations",
        headers=headers,
        params={"classification_id": f"eq.{classification_id}"}
    )
    return response.json()

def create_classification(department_id, name, color="#6366F1"):
    """Create a new classification"""
    response = httpx.post(
        f"{SUPABASE_URL}/rest/v1/warehouse_classifications",
        headers=headers,
        json={
            "department_id": department_id,
            "name": name,
            "color": color
        }
    )
    return response.json()

def create_location(department_id, classification_id, name, parent_id=None):
    """Create a new location/folder"""
    data = {
        "department_id": department_id,
        "classification_id": classification_id,
        "name": name
    }
    if parent_id:
        data["parent_id"] = parent_id
    
    response = httpx.post(
        f"{SUPABASE_URL}/rest/v1/warehouse_locations",
        headers=headers,
        json=data
    )
    return response.json()

def generate_item_number():
    """Generate a unique item number"""
    return f"IT-{datetime.now().strftime('%Y%m%d')}-{str(uuid.uuid4())[:8].upper()}"

def insert_inventory_item(item_data):
    """Insert a single inventory item"""
    response = httpx.post(
        f"{SUPABASE_URL}/rest/v1/inventory_items",
        headers=headers,
        json=item_data
    )
    if response.status_code >= 400:
        print(f"Error inserting {item_data['item_name']}: {response.text}")
        return None
    return response.json()

def main():
    # Read Excel file
    excel_path = os.path.join(os.path.dirname(__file__), "..", "2026_01_08_10_33_25_tovars_IT equipment and devices.xls")
    df = pd.read_excel(excel_path)
    
    print(f"Found {len(df)} items in Excel file")
    print("Columns:", df.columns.tolist())
    print()
    
    # Get departments and find IT department
    departments = get_departments()
    print("Available departments:")
    for dept in departments:
        print(f"  - {dept['name']} (code: {dept['code']}, id: {dept['id']})")
    
    # Find IT department (look for IT, Information Technology, etc.)
    it_dept = None
    for dept in departments:
        if 'IT' in dept['code'].upper() or 'IT' in dept['name'].upper() or 'INFORMATION' in dept['name'].upper():
            it_dept = dept
            break
    
    if not it_dept:
        print("\nNo IT department found. Creating one...")
        # We can't create department without admin access, so let's use the first available or ask
        print("Please specify which department to use (enter the code):")
        return
    
    print(f"\nUsing department: {it_dept['name']} ({it_dept['id']})")
    
    # Get classifications for this department
    classifications = get_classifications(it_dept['id'])
    print(f"\nExisting classifications for {it_dept['name']}:")
    for cls in classifications:
        print(f"  - {cls['name']} (id: {cls['id']})")
    
    # Find or create "IT Equipment and Devices" classification
    target_classification = None
    for cls in classifications:
        if 'equipment' in cls['name'].lower() or 'device' in cls['name'].lower():
            target_classification = cls
            break
    
    if not target_classification and classifications:
        # Use first available classification
        target_classification = classifications[0]
        print(f"\nUsing classification: {target_classification['name']}")
    elif not target_classification:
        print("\nNo classification found. Creating 'IT Equipment and Devices'...")
        result = create_classification(it_dept['id'], "IT Equipment and Devices", "#3B82F6")
        if result:
            target_classification = result[0] if isinstance(result, list) else result
            print(f"Created classification: {target_classification}")
    
    if not target_classification:
        print("Error: Could not find or create classification")
        return
    
    # Get locations for this classification
    locations = get_locations(target_classification['id'])
    print(f"\nExisting locations in {target_classification['name']}:")
    for loc in locations:
        print(f"  - {loc['name']} (id: {loc['id']})")
    
    # Find or create a location for the items
    target_location = None
    for loc in locations:
        # Find any location to put items in
        target_location = loc
        break
    
    if not target_location:
        print("\nNo location found. Creating 'Main Storage'...")
        result = create_location(it_dept['id'], target_classification['id'], "Main Storage")
        if result:
            target_location = result[0] if isinstance(result, list) else result
            print(f"Created location: {target_location}")
    
    if not target_location:
        print("Error: Could not find or create location")
        return
    
    print(f"\n{'='*60}")
    print(f"Importing items to:")
    print(f"  Department: {it_dept['name']}")
    print(f"  Classification: {target_classification['name']}")
    print(f"  Location: {target_location['name']}")
    print(f"{'='*60}\n")
    
    # Import items
    success_count = 0
    error_count = 0
    
    for index, row in df.iterrows():
        item_name = str(row['Item name']).strip() if pd.notna(row['Item name']) else ""
        quantity = int(row['Quantity']) if pd.notna(row['Quantity']) else 0
        min_quantity = int(row['Min.Quantity']) if pd.notna(row['Min.Quantity']) else 0
        unit = str(row['Unit of Measure']).strip() if pd.notna(row['Unit of Measure']) and str(row['Unit of Measure']) != 'nan' else "pcs"
        
        if not item_name:
            continue
        
        item_data = {
            "department_id": it_dept['id'],
            "classification_id": target_classification['id'],
            "location_id": target_location['id'],
            "item_name": item_name,
            "item_number": generate_item_number(),
            "quantity": quantity,
            "min_quantity": min_quantity,
            "unit": unit if unit != 'nan' else "pcs"
        }
        
        result = insert_inventory_item(item_data)
        if result:
            success_count += 1
            print(f"✓ Imported: {item_name} (qty: {quantity})")
        else:
            error_count += 1
            print(f"✗ Failed: {item_name}")
    
    print(f"\n{'='*60}")
    print(f"Import completed!")
    print(f"  Success: {success_count}")
    print(f"  Errors: {error_count}")
    print(f"{'='*60}")

if __name__ == "__main__":
    main()

File: scripts/import_inventory.ts
// Script to import inventory items from Excel file to Supabase
// Run with: npx tsx scripts/import_inventory.ts
// Requires: EMAIL and PASSWORD environment variables for authentication

import { createClient } from '@supabase/supabase-js';
import XLSX from 'xlsx';
import path from 'path';
import fs from 'fs';
import { fileURLToPath } from 'url';
import * as readline from 'readline';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

const SUPABASE_URL = 'https://edumcnnilpnbdxcjpchw.supabase.co';
const SUPABASE_KEY = 'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJzdXBhYmFzZSIsInJlZiI6ImVkdW1jbm5pbHBuYmR4Y2pwY2h3Iiwicm9sZSI6ImFub24iLCJpYXQiOjE3NjczMzI1ODEsImV4cCI6MjA4MjkwODU4MX0.nRPDvserHqLnNx78UArKHuzJp5J4C0-FT9noHtmXfFU';

const supabase = createClient(SUPABASE_URL, SUPABASE_KEY);

interface ExcelRow {
  'Item name': string;
  'Quantity': number;
  'Unit of Measure'?: string;
  'Min.Quantity': number;
}

function generateItemNumber(): string {
  const date = new Date().toISOString().slice(0, 10).replace(/-/g, '');
  const random = Math.random().toString(36).substring(2, 10).toUpperCase();
  return `IT-${date}-${random}`;
}

function prompt(question: string): Promise<string> {
  const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout
  });
  
  return new Promise((resolve) => {
    rl.question(question, (answer) => {
      rl.close();
      resolve(answer);
    });
  });
}

async function main() {
  console.log('='.repeat(60));
  console.log('Inventory Import Script');
  console.log('='.repeat(60));
  console.log('\nThis script requires authentication to access the database.');
  console.log('Please enter your Supabase credentials:\n');

  // Get credentials
  const email = process.env.EMAIL || await prompt('Email: ');
  const password = process.env.PASSWORD || await prompt('Password: ');

  // Sign in
  console.log('\nSigning in...');
  const { data: authData, error: authError } = await supabase.auth.signInWithPassword({
    email,
    password
  });

  if (authError) {
    console.error('Authentication failed:', authError.message);
    return;
  }

  console.log(`✓ Signed in as: ${authData.user?.email}\n`);

  console.log('Starting inventory import...\n');

  // Read Excel file
  const excelPath = path.join(__dirname, '..', '2026_01_08_10_33_25_tovars_IT equipment and devices.xls');
  
  if (!fs.existsSync(excelPath)) {
    console.error(`Excel file not found: ${excelPath}`);
    return;
  }

  const workbook = XLSX.readFile(excelPath);
  const sheetName = workbook.SheetNames[0];
  const worksheet = workbook.Sheets[sheetName];
  const data: ExcelRow[] = XLSX.utils.sheet_to_json(worksheet);

  console.log(`Found ${data.length} items in Excel file\n`);

  // Get all departments
  const { data: departments, error: deptError } = await supabase
    .from('departments')
    .select('*');

  if (deptError) {
    console.error('Error fetching departments:', deptError.message);
    return;
  }

  console.log('Available departments:');
  departments?.forEach(dept => {
    console.log(`  - ${dept.name} (code: ${dept.code}, id: ${dept.id})`);
  });

  // Find Warehouse department
  const warehouseDept = departments?.find(d => 
    d.code === 'WH' || d.code === 'WAREHOUSE' || d.name.toLowerCase().includes('warehouse')
  );

  if (!warehouseDept) {
    console.error('\nWarehouse department not found!');
    console.log('Available department codes:', departments?.map(d => d.code));
    return;
  }

  console.log(`\nUsing department: ${warehouseDept.name} (${warehouseDept.id})`);

  // Get classifications for Warehouse department
  const { data: classifications, error: classError } = await supabase
    .from('warehouse_classifications')
    .select('*')
    .eq('department_id', warehouseDept.id);

  if (classError) {
    console.error('Error fetching classifications:', classError.message);
    return;
  }

  console.log('\nAvailable classifications:');
  classifications?.forEach(cls => {
    console.log(`  - ${cls.name} (id: ${cls.id})`);
  });

  // Find "IT Equipments" classification
  const targetClassification = classifications?.find(c => 
    c.name.toLowerCase().includes('it equipment') || 
    c.name.toLowerCase().includes('it equipments')
  );

  if (!targetClassification) {
    console.error('\nIT Equipment classification not found!');
    return;
  }

  console.log(`\nUsing classification: ${targetClassification.name} (${targetClassification.id})`);

  // Get locations for this classification
  const { data: locations, error: locError } = await supabase
    .from('warehouse_locations')
    .select('*')
    .eq('classification_id', targetClassification.id)
    .is('parent_id', null);

  if (locError) {
    console.error('Error fetching locations:', locError.message);
    return;
  }

  console.log('\nAvailable locations:');
  locations?.forEach(loc => {
    console.log(`  - ${loc.name} (id: ${loc.id})`);
  });

  let targetLocation = locations?.[0];

  // Create a location if none exists
  if (!targetLocation) {
    console.log('\nNo location found. Creating "Main Storage"...');
    const { data: newLoc, error: createLocError } = await supabase
      .from('warehouse_locations')
      .insert({
        department_id: warehouseDept.id,
        classification_id: targetClassification.id,
        name: 'Main Storage'
      })
      .select()
      .single();

    if (createLocError) {
      console.error('Error creating location:', createLocError.message);
      return;
    }
    targetLocation = newLoc;
    console.log(`Created location: ${targetLocation.name}`);
  }

  console.log(`\nUsing location: ${targetLocation.name} (${targetLocation.id})`);

  console.log('\n' + '='.repeat(60));
  console.log('Starting item import...');
  console.log('='.repeat(60) + '\n');

  // Import items
  let successCount = 0;
  let errorCount = 0;

  for (const row of data) {
    const itemName = row['Item name']?.toString().trim();
    if (!itemName) continue;

    const quantity = Number(row['Quantity']) || 0;
    const minQuantity = Number(row['Min.Quantity']) || 0;
    const unit = row['Unit of Measure']?.toString().trim() || 'pcs';

    const itemData = {
      department_id: warehouseDept.id,
      classification_id: targetClassification.id,
      location_id: targetLocation.id,
      item_name: itemName,
      item_number: generateItemNumber(),
      quantity: quantity,
      min_quantity: minQuantity,
      unit: unit === 'nan' || unit === '' ? 'pcs' : unit
    };

    const { data: inserted, error: insertError } = await supabase
      .from('inventory_items')
      .insert(itemData)
      .select()
      .single();

    if (insertError) {
      console.log(`✗ Failed: ${itemName} - ${insertError.message}`);
      errorCount++;
    } else {
      console.log(`✓ Imported: ${itemName} (qty: ${quantity})`);
      successCount++;
    }
  }

  console.log('\n' + '='.repeat(60));
  console.log('Import completed!');
  console.log(`  Success: ${successCount}`);
  console.log(`  Errors: ${errorCount}`);
  console.log('='.repeat(60));
}

main().catch(console.error);
